source/arm9/math.c

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: Zlib
+//
+// Copyright (C) 2024 Dominik Kurz
+#include <nds/arm9/math.h>
+
+
+float hw_sqrtf(float x)
+{
+    union{float f; u32 i;}xu;
+    xu.f = x;
+    u32 mantissa32 = xu.i & ((1 << 23) - 1); // extracts mantissa via bitmask
+    mantissa32 += (1 << 23); // adds implicit bit to mantissa.
+    // If you were planning to handle subnormal numbers 
+    // you would check if the number is subnormal before 
+    // adding the implicit bit.
+    // You would also need to do the shift of the mantissa and exponent
+    // differently, probably should use __builtin_clz for this
+    // as the count leading zeroes (clz) instruction is 1 cycle on arm
+    // but this is not my problem
+    // performance-critical code has subnormals turned off
+    // due to -ffast-math 
+    u64 mantissa = mantissa32; // cast so that the shift doesnt go wrong 
+    mantissa <<= (((xu.i & (1 << 23)) == 0) + 25);
+    // applies additional shift depending on whether exponent is even or odd
+    // 25 is 23+2, the +2 is necessary for rounding
+    REG_SQRTCNT = SQRT_64;  
+    REG_SQRT_PARAM = mantissa;
+    // starts hardware squareroot 
+    // It is critical that this happens as early as possible so that
+    // we have time to do other stuff in the meantime
+    u32 raw_exponent = (xu.i & (0xff << 23));
+    u32 sign = (xu.i & (1 << 31));
+    // check if exponent is 0
+    if (raw_exponent != 0)  
+    {
+        if (sign) //check if negative
+        {
+        // Expected behavior:
+        // sqrt(-f)=+qNaN, sqrt(-NaN)=+qNaN,sqrt(-Inf)=+qNaN 
+        // This branch is probably not very useful.
+        // If someone is taking the sqrt of a negative value 
+        // you will usually call abs() on the argument of the input
+        // so really, you want negatives to be treated like positives
+        // also makes the logic unnecessarily complicated
+        // and computing an absolute value is more expensive 
+        // than simply ignoring the sign of a float
+        // Supposed to be a signalling NaN (sNaN)
+        // Honestly there is probably a better way to do this.
+        // as the IEEE 754 standard doesnt specify how signalling NaN
+        // is encoded. Hard float arm stuff does it with a 1 in the least
+        // significant bit and zeros in the rest of the mantissa
+        // Quiet NaNs  on ARM are a 1 in both 
+        // the most and least significant bit of the mantissa.
+        // Example:
+        // Quiet NaN mantissa (100000.....00001)
+        // Signalling NaN mantissa (00000.....00001)
+            xu.i= (255 << 23) |( ( 1 << 22 ) |1 );
```

Comment:
Hardware ARM implementations may leave a 1 in the LSB, but the software implementation of libm leaves a 0 there... I'm not sure what's the right behaviour here.

Reply:
Probably not important as long as it is a type of NaN. It's mostly to help people find bugs I think, returning NaN isn't that useful beyond that.

Reply:
Sure.

In any case, the `sqrtf()` of `math.h` returns a quiet NaN (`got` is the `hw_sqrt` version, `expected` is the `math.h` version):

![image](https://github.com/blocksds/libnds/assets/6632664/386ebf3f-5473-4f1a-a8ab-df84ad7a6acd)

Maybe it's better for them to be the same?

Reply:
You're testing a sNaN as input, so it preserves it.
The branch you quoted is for negative numbers, in which case they are the both quiet NaNs.
![image](https://github.com/blocksds/libnds/assets/47481645/ab453b95-1616-409c-8fe9-29f808d8accb)

Reply:
Actually nevermind, I can probably just delete the bit that makes a NaN signalling.

Reply:
> You're testing a sNaN as input, so it preserves it. The branch you quoted is for negative numbers, in which case they should both be quiet NaNs.

According to this: https://en.wikipedia.org/wiki/NaN

> Signaling NaNs, or sNaNs, are special forms of a NaN that, when consumed by most operations, should raise the invalid operation exception and then, if appropriate, be "quieted" into a qNaN that may then propagate.

So it looks like `math.h` is doing the right thing by turning the sNaN into a qNaN.